### PR TITLE
metadata: import, export, create csv data files

### PR DESCRIPTION
**Pull request by rerowep**
* Implements import and export of resources in csv format.
* Implements creation of csv files from json files.

Co-Authored-by: Peter Weber <peter.weber@rero.ch>

## Why are you opening this PR?

- Which task/US does it implement?
- Which issue does it fix?

## Dependencies

My PR depends on the following `rero-ils-ui`'s PR(s):

* rero/rero-ils-ui#<xx>

## How to test?

- What command should I have to run to test your PR?
- What should I test through the UI?

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Cypress tests successful?

### CODE REVIEW

**Review by iGormilhit — COMMENTED** [review]
Commit message title: `metadata: import, export, create csv data files`

File: rero_ils/modules/cli.py [hunk 1]
```diff
@@ -1178,6 +1212,49 @@ def reindex(pid_type, no_info):
         )
 
 
+@utils.command('reindex_missing')
+@click.option('-t', '--pid-type', multiple=True, required=True)
+@click.option('-v', '--verbose', 'verbose', is_flag=True, default=False)
+@with_appcontext
+def reindex_missing(pid_type, verbose):
```
Comment (zannkukai) [inline rero_ils/modules/cli.py:1219]: cosmetic : maybe `pid_types` name would be better if this argument is an array.

File: rero_ils/modules/cli.py [hunk 2]
```diff
@@ -1178,6 +1212,49 @@ def reindex(pid_type, no_info):
         )
 
 
+@utils.command('reindex_missing')
+@click.option('-t', '--pid-type', multiple=True, required=True)
```
Comment (zannkukai) [inline rero_ils/modules/cli.py:1216]: cosmetic : `@click.option('-t', '--pid-type', 'pid-types', multiple=True, required=True)`
See next comment

File: rero_ils/modules/cli.py [hunk 3]
```diff
@@ -1782,3 +1859,267 @@ def add_cover_urls(verbose):
...
+        file_errors.close()
+        if errors_count == 0:
+            os.remove(file_name_errors)
+    click.secho(
+        'Created: {count} Errors: {error_count}'.format(
+            count=count-errors_count,
+            error_count=errors_count
+        ),
+        fg='yellow'
+    )
```
Comment (zannkukai) [inline rero_ils/modules/cli.py:1989]: bad identation ? this `if` statement could/should be outside of the `with open(...) as infile` statement.

The `click.secho` statement could/should be the `else` statement of the same if (display the message only if errors are detected)

File: rero_ils/modules/cli.py [hunk 4]
```diff
@@ -1782,3 +1859,267 @@ def add_cover_urls(verbose):
...
+    click.secho('  Load metatada: {file_name}'.format(
+        file_name=file_name_metadata
+    ))
+    bulk_load_metadata(pid_type=record_type, metadata=file_name_metadata,
+                       bulk_count=bulk_count, verbose=verbose, reindex=reindex)
+
+
+@fixtures.command('bulk_save')
+@click.argument('output_directory')
+@click.option('-t', '--pid_type', 'pid_type', multiple=True, default=['all'])
```
Comment (zannkukai) [inline rero_ils/modules/cli.py:2047]: cosmetic : same remarks about argument name

File: tests/utils.py [hunk 5]
```diff
@@ -350,3 +350,15 @@ def item_record_to_a_specific_loan_state(
 
     assert loan['state'] == loan_state
     return item, loan
+
+
+def cmp_files(file1_name, file2_name):
+    """Compair files."""
+    with open(file1_name) as file1:
+        with open(file2_name) as file2:
```
Comment (zannkukai) [inline tests/utils.py:358]: Could be merge on a single line :
```python
with open("file_1.txt") as f1, open("file_2.txt") as f2:
       line_file1 = f1.readline()
       line_file2 = f2.readline()
```

**Review by BadrAly — APPROVED** [review]

File: rero_ils/modules/cli.py (hunk 1 above)
Comment (rerowep) [reply rero_ils/modules/cli.py:1219]: ok

File: rero_ils/modules/cli.py (hunk 3 above)
Comment (rerowep) [reply rero_ils/modules/cli.py:1989]: ok

File: tests/utils.py (hunk 5 above)
Comment (rerowep) [reply tests/utils.py:372]: ok

**Review by zannkukai — APPROVED** [review]

File: rero_ils/modules/api.py [hunk 6]
```diff
@@ -441,6 +441,13 @@ def organisation_pid(self):
         """Get organisation pid for circulation policy."""
         return extracted_data_from_ref(self.get('organisation'))
 
+    @classmethod
+    def get_metadata_identifier_names(cls):
+        """Get metadata and identif table names."""
```
Comment (lauren-d) [inline rero_ils/modules/api.py:446]: typo in identif => identifier

**Review by lauren-d — APPROVED** [review]

**Review by zannkukai — APPROVED** [review]

**Review by iGormilhit — APPROVED** [review]

**Pull request merged**